src/main/java/com/gojek/beast/sink/bq/BqSink.java

```diff
@@ -29,48 +29,93 @@
         //if bq has errors
         if (response.hasErrors()) {
             //parse the error records
-            Map<Record, List<BQInsertionRecordsErrorType>> parsedRecords = responseParser.parseBQResponse(records, response);
-            //sink error records
-            boolean isTheSinkSuccessful = errorHandler.handleErrorRecords(parsedRecords).isSuccess();
-            if (!isTheSinkSuccessful) {
-                return new InsertStatus(isTheSinkSuccessful, response.getInsertErrors());
+            Map<Record, List<BQRecordsErrorType>> parsedRecords = responseParser.parseBQResponse(validRecordList, response);
+
+            // if batch contains records that we can't really handle, fail whole batch
+            List<Record> unhandledRecords = unhandledInsertsFilter.filter(parsedRecords);
+            if(!unhandledRecords.isEmpty()) {
+                log.info("Batch with records size: {} contains invalid records, marking this batch to fail", unhandledRecords.size());
+                statsClient.gauge("data.error.records,type=invalid", unhandledRecords.size());
+                return new InsertStatus(false, response.getInsertErrors());
             }
-            Records bqValidRecords = errorHandler.getBQValidRecords(parsedRecords);
-            if (!bqValidRecords.getRecords().isEmpty()) { // there are valid records
-                //insert the valid records into bq
-                isTheSinkSuccessful &= !insertIntoBQ(bqValidRecords).hasErrors();
+
+            // retryable records
+            List<Record> retryableRecords = retryableInsertsFilter.filter(parsedRecords);
+            if (!retryableRecords.isEmpty()) {
+                // try inserting valid records into bq
+                InsertAllResponse retriedResponse = insertIntoBQ(retryableRecords);
+                if(retriedResponse.hasErrors()) {
+                    return new InsertStatus(true, retriedResponse.getInsertErrors());
+                }
+            }
+
+            // DLQ sinkable records
+            List<Record> oobRecords = oobInsertsFilter.filter(parsedRecords);
+            if (!oobRecords.isEmpty()) {
+                log.info("Error handler parsed OOB records size {}, handoff to the writer {}", oobRecords.size(), errorWriter.getClass().getSimpleName());
+                statsClient.gauge("data.error.records,type=oob", oobRecords.size());
+                List<Record> errList = dlqSinkableRecords.getOrDefault(BQRecordsErrorType.OOB, new ArrayList<>());
+                errList.addAll(oobRecords);
+                dlqSinkableRecords.put(BQRecordsErrorType.OOB, errList);
             }
-            return new InsertStatus(isTheSinkSuccessful, response.getInsertErrors());
         }
-        // return the original response
-        return new InsertStatus(!response.hasErrors(), response.getInsertErrors());
+
+        // sink empty records to DLQ
+        if (!dlqSinkableRecords.isEmpty()) {
```

Comment:
Any reason why we don't insert into dlq when the deserialisation failed in the ConsumerRecordConverter. We should compose the ConsumerRecordCoverter with DLQ Handler and insert the failed records and proceed processing the other records. The BQ Sink push which gets the `Records` batch should only contain rest of the records.

BQSink is way down the

Reply:
We can't really do this in `ConsumerRecordConverter` as it's single-threaded whereas BQSink will have multiple instances running in parallel.

Reply:
The no of., of times we've encountered this was may be 5 times in the last couple of months. I think it's safer to shove the erroneous bytes into dlq in the consumer thread, since we are not going to process millions of corrupt records. If we do then we have a larger problem to sort than the design.
Also, I think doing it in ConsumerRecordConverter is lot cleaner as the notion of failing deseralization is only with the converter and don't need to pass the info to the sinks.